File: 3.nsys/cfdcode-onegpu/solution/C/4.acc-nodata/jacobi.c
#include <stdio.h>
#include <nvToolsExt.h>
#include "jacobi.h"

void jacobistep_acc(double **psinew, double **psi, int m, int n)
{
  int i, j;
  //nvtxRangeId_t jacc_Id = nvtxRangeStartA("jacobi_acc");
#pragma acc parallel loop collapse(2)
  for(i=1;i<=m;i++)
    {
      for(j=1;j<=n;j++)
	{
	  psinew[i][j]=0.25*(psi[i-1][j]+psi[i+1][j]+psi[i][j-1]+psi[i][j+1]);
        }
    }
  //nvtxRangeEnd(jacc_Id);
}


double deltasq(double **newarr, double **oldarr, int m, int n)
{
  int i, j;

  double dsq=0.0;
  double tmp;

  for(i=1;i<=m;i++)
    {
      for(j=1;j<=n;j++)
	{
	  tmp = newarr[i][j]-oldarr[i][j];
	  dsq += tmp*tmp;
        }
    }

  return dsq;
}

File: 3.nsys/cfdcode-onegpu/solution/Fortran/4.no-acc-data/jacobi.f90
module jacobi
  USE nvtx
  implicit none

contains

subroutine jacobistep_acc(psinew, psi, m, n)
  integer :: m, n
  integer :: i, j

  double precision, dimension(0:m+1, 0:n+1) :: psinew, psi
 
   call nvtxStartRange("jacobi_acc",1) 
   !$acc parallel loop collapse(2)
   do j = 1, n
      do i = 1, m
          psinew(i, j) = 0.25d0*(psi(i+1, j) + psi(i-1, j) + &
                             psi(i, j+1) + psi(i, j-1)     )
      end do
  end do
  !$acc end parallel loop
  call nvtxEndRange()

end subroutine jacobistep_acc

double precision function deltasq(new, old, m, n)

  integer :: m, n, i, j
  double precision, dimension(0:m+1, 0:n+1) :: new, old
  double precision :: deltasq

  integer :: ierr
   
   deltasq = 0.d0
     do j = 1, n
      do i = 1, m
      deltasq = deltasq + (new(i,j)-old(i,j))**2
     end do
    end do

end function deltasq

end module jacobi
                                    


